lib/rex/proto/ldap/server.rb

```diff
@@ -0,0 +1,304 @@
+# -*- coding: binary -*-
+
+require 'rex/socket'
+require 'net/ldap'
+
+module Rex
+module Proto
+module LDAP
+
+class Server
+
+  module LdapClient
+    attr_accessor :authenticated
+    #
+    # Initialize LDAP client state
+    #
+    def init_ldap_client
+      self.authenticated = false
+    end
+  end
+
+  class MockLdapClient
```

Comment:
What's the intent of `MockLdapClient`? 👀 

Looks like it's a similar pattern to previous dns server:
https://github.com/rapid7/metasploit-framework/blob/6175e389712fd8e5ab3948d0f03414d887cd5193/lib/rex/proto/dns/server.rb#L140

But `mock_ldap_client?` and `mock_dns_client?` are never called, or maybe I'm missing out on some important magic in framework 😄

Reply:
Not in upstream code, no. These came about when i was doing some DNS work and needed to differentiate quickly/easily - the LDAP piece just got its own version. Can pull 'em if you want, though i dont see the damage per se.
The actual Module is critical: thats the pattern to unify TCP/UDP split-stack services otherwise you have different interfaces for your UDP client and your TCP client and need to write differentiated `write` handlers and such for each case.

Reply:
@adfoster-r7: does the implementation of the client objects make sense from the perspective or `:read`/`:write` abstraction (can i mark this resolved)?